### PR TITLE
[ricos-common] feat(theme): customStyles API + refactoring

### PR DESCRIPTION
**Pull request by Barackos**
### :rocket: New Feature
- theme API features - [Check it out](https://wix-incubator.github.io/rich-content/docs/ricos/theming)
    - typography API - fontFamily
    - custom fields addition
- Detailed `RicosTheme` Documentation

### 🏠 Internal
- themeStrategy refactor, automated css-vars generation, theme unit tests

### CODE REVIEW

File: packages/common/web/statics/styles/fonts.scss [hunk 1]
```diff
@@ -1,4 +1,52 @@
...
+  font-weight: var(--ricos-custom-#{$name}-font-weight);
+  text-decoration: var(--ricos-custom-#{$name}-text-decoration, none);
+  // TODO: this 'if' should be removed. Removing currently breaks blog theme.
+  @if $name != 'link' and $name != 'hashtag' {
+    line-height: var(--ricos-custom-#{$name}-line-height, map-get($text-heights, $name));
+  }
+  // TODO: this 'if' should be removed. Removing currently breaks blog theme.
+  @if $isViewerHeader and ($name == 'h1' or $name == 'h2' or $name == 'h3') {
+    min-height: var(--ricos-custom-#{$name}-min-height, map-get($text-heights, $name));
+  }
```
Comment (Barackos) [inline packages/common/web/statics/styles/fonts.scss:48]: @lxgreen Dilemma:
Blog theme breaks if I don't add those `if`s, but it limits our API.
`link`, `hashtag`, `h1`-`h3` originally don't implement a default value for those css properties, so these `if`s prevents setting a value to keep them "untouched", but it harms our API (`line-height` is non-customisable for `link`, for example).

File: packages/ricos-common/web/src/themeStrategy/generators/customs.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { CssVarsObject, RicosCustomTheme, CustomTextualStyle } from '../themeTypes';
+
+const toEntries = (customs: RicosCustomTheme): [string, CustomTextualStyle][] =>
+  Object.entries(customs);
+
+const toVars = (customs: RicosCustomTheme) =>
+  toEntries(customs).reduce(
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/generators/customs.ts:7]:
```suggestion
  Object.entries(customs).reduce(
```

File: packages/ricos-common/web/src/themeStrategy/generators/customs.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { CssVarsObject, RicosCustomTheme, CustomTextualStyle } from '../themeTypes';
+
+const toEntries = (customs: RicosCustomTheme): [string, CustomTextualStyle][] =>
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/generators/customs.ts:3]: one-liner used in a single place -- smells like redundant complication

File: packages/ricos-common/web/src/themeStrategy/generators/customs.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  Object.entries(customs);
+
+const toVars = (customs: RicosCustomTheme) =>
+  toEntries(customs).reduce(
+    (prev, [fieldName, customStyle]) => ({
+      ...prev,
+      [`custom-${fieldName.toLowerCase()}FontFamily`]: customStyle.fontFamily,
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/generators/customs.ts:10]: how about to reduce it even more by iterating on `Object.entries(customStyle)`:
```
[`custom-${fieldName.toLowerCase()}${toTitleCase(customStyleProp)}`]: customStyle[customStyleProp]
```

File: packages/ricos-common/web/src/themeStrategy/generators/customs.spec.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { CssVarsObject, RicosCustomTheme } from '../themeTypes';
+import { mockCustoms, mockOutput } from '../../../tests/mockCustoms';
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/generators/customs.spec.ts:2]: technically, these are not mocks, just test cases

File: packages/ricos-common/web/src/themeStrategy/generators/palette.spec.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    backgroundColor: '#0E092B',
+    backgroundColorTuple: '14, 9, 43',
+  };
+
+  it('should return empty colors object', () => {
+    const cssVars = createPalette();
+    expect(cssVars).toEqual({});
+  });
+
+  it('should throw if theme is unknwon', () => {
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/generators/palette.spec.ts:22]:
```suggestion
  it('should throw if theme is unknown', () => {
```

File: packages/ricos-common/web/src/themeStrategy/themeUtils.ts [hunk 7]
```diff
@@ -79,3 +80,23 @@ export function toCssRgbA(hexColor: string, opacity: number): string {
   }
   throw new Error('Bad Hex');
 }
+
+export const toDashedKey = (str: string) =>
+  str.replace(/([A-Z])/g, (all, letter) => '-' + letter.toLowerCase());
+
+const spacing = '    ';
```
Comment (lxgreen) [inline packages/ricos-common/web/src/themeStrategy/themeUtils.ts:87]:
```suggestion
const spacing = ' '.repeat(4); // less error-prone
```

**Review by lxgreen — APPROVED** [review]

File: docs/docs/ricos/Theming.mdx [hunk 8]
```diff
@@ -4,4 +4,128 @@ title: Theming
 sidebar_label: Theming
 ---
 
-TODO
+```jsx
```
[the remark this answers is not in the document]
Comment (Barackos) [reply docs/docs/ricos/Theming.mdx:7]: I'll convert them to a table instead.
This is to let consumers know what the prop fields are, following our prop-fields API convention ([example](https://wix-incubator.github.io/rich-content/docs/ricos/ricos-api/#linksettings)).

File: packages/common/web/statics/styles/fonts.scss [hunk 9]
```diff
@@ -1,4 +1,52 @@
...
+  @else if $name == 'link' {
+    color: $ricos-link-color;
+  }
+  font-family: var(--ricos-custom-#{$name}-font-family);
+  font-size: var(--ricos-custom-#{$name}-font-size, map-get($font-sizes, $name));
+  font-style: var(--ricos-custom-#{$name}-font-style);
+  font-weight: var(--ricos-custom-#{$name}-font-weight);
+  text-decoration: var(--ricos-custom-#{$name}-text-decoration, none);
+  // TODO: this 'if' should be removed. Removing currently breaks blog theme.
```
[the remark this answers is not in the document]
Comment (Barackos) [reply packages/common/web/statics/styles/fonts.scss:41]: True.
Removing the 2nd stage "wixTypography" as it has been removed, might be considered in future phase.

File: docs/docs/ricos/Theming.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (lxgreen) [reply docs/docs/ricos/Theming.mdx]: well, this convention is pre-typescript as well, it's not mandatory to follow it forever. i leave it for you to decide.

**Pull request merged**